client/client.go

```diff
@@ -120,18 +120,24 @@ func (c *LogClient) waitForRootUpdate(ctx context.Context) error {
 	}
 	startTreeSize := c.root.TreeSize
 	for i := 0; ; i++ {
-		if err := c.UpdateRoot(ctx); err != nil {
+		err := c.UpdateRoot(ctx)
+		switch x := status.Code(err); x {
+		case codes.OK:
+			if c.root.TreeSize > startTreeSize {
+				return nil
+			}
+		case codes.Unavailable: // Retry.
+		default:
 			return err
 		}
-		if c.root.TreeSize > startTreeSize {
-			return nil
-		}
-		if err := ctx.Err(); err != nil {
+
+		select {
+		case <-ctx.Done():
 			return status.Errorf(codes.DeadlineExceeded,
-				"%v. TreeSize: %v, want > %v. Tried %v times.",
-				err, c.root.TreeSize, startTreeSize, i+1)
+				"%v. TreeSize: %v, want > %v. Tried %v times: %v",
+				err, c.root.TreeSize, startTreeSize, i+1, ctx.Err())
+		case <-time.After(b.Duration()):
```

Comment:
Brain melting - what does this actually do? Do we get a new channel created when we enter the select{}, even if `<-ctx.Done()` is selected?

Might be clearer to write it with a `time.Ticker`:
```golang
tick := time.NewTicker(b.Duration())
...
select {
  ...
  case <-tick.C
}
```

Reply:
Fair question :-) 

`time.After` is basically on a one time ticker. It's super useful because it's a channel and we can put it into that select block.  The select block allows us to terminate early if the context is canceled or times out first rather than needing to wait for the full duration only to discover that we had timed out :-) 

Yes, the channel gets created when we enter the select.  This is the use case it's designed for. See https://golang.org/pkg/time/#After. https://golang.org/ref/spec#Select_statements

Reply:
Okaaaay, but it's evil witchcraft to me :)